### PR TITLE
Support multipart form data

### PR DESCRIPTION
**Pull request by goodexpert**
See #1873

## Checklist

- [o] `npm test` passes on your machine
- [o] New tests added or existing tests modified to cover all changes
- [o] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

### CODE REVIEW

**Comment by raymondfeng** [comment]
@goodexpert What's take on this alternative solution - https://github.com/strongloop/loopback-next/pull/1936/commits/98f6caf891f84983a4bc3f0d8165a9be096497d9?

**Comment by goodexpert** [comment]
> @goodexpert What's take on this alternative solution - [98f6caf](https://github.com/strongloop/loopback-next/commit/98f6caf891f84983a4bc3f0d8165a9be096497d9)?

I have submitted pull request on last month. But I did not check listing and formatting issue.

**Comment by raymondfeng** [comment]
@goodexpert Thank you for the contribution. 

From the process perspective, you don't have to open a new PR. You can keep pushing changes into your repository and the PR submitted from that repo will be automatically refreshed by github.

File: packages/rest/src/parser.ts [hunk 1]
```diff
@@ -56,6 +57,34 @@ const parseFormBody: (
...
+      for (const key of Object.keys(fields)) {
+        const value = fields[key];
+
+        if (value instanceof Array) {
+          json[key] = value[0];
+        } else {
+          json[key] = value;
+        }
+      }
+      resolve({...json, files: files});
```
Comment (raymondfeng) [inline packages/rest/src/parser.ts]: How do we access the `part` as streams?

**Comment by bajtos** [comment]
+1 for not keeping the work in the same pull request. @goodexpert if you are new to GitHub pull request flow, then please read the following article for step-by-step guide explaining how to work with `git` when contributing pull requests. The important part is that you want to periodically rebase your feature branch on top of the latest upstream master. 

As for this particular pull request, I am personally reluctant to add file upload as the built-in feature of `@loopback/rest` for the reasons outlined in https://github.com/strongloop/loopback-next/issues/1873#issuecomment-432974060:

_Having file-uploads outside of `@loopback/rest` will greatly benefit people NOT using this feature. The size of their dependencies will be smaller and as an important consequence, their applications will have less sources of potential security vulnerabilities._

What I would like to do instead:
 - Inside `@loopback/rest`, implement an extension point allowing 3rd party extensions to provide parsers for different content types (e.g. multi-part).
 - Create a new LB extension providing multi-part parser (file upload support). I think loopback-next should provide an official/reference implementation of this extension inside loopback-next monorepo, but we should also make it clear that we are welcoming community efforts to provide alternative implementations.

@strongloop/loopback-next thoughts?

**Comment by raymondfeng** [comment]
See https://github.com/strongloop/loopback-next/tree/master/packages/rest/test/acceptance/file-upload

**Comment by bajtos** [comment]
I am closing this pull request as superseded by #1936.

@goodexpert if you would like to contribute an extension providing file-upload body parser, then please open a new pull request.

**Pull request closed without merging**